doc/developers/contributing.rst

```diff
@@ -268,100 +249,115 @@ If any of the above seems like magic to you, then look up the `Git documentation
-  the example. For more details on writing and building the
-  documentation, see the :ref:`contribute_documentation` section.
-
-* The documentation should also include expected time and space complexity
-  of the algorithm and scalability, e.g. "this algorithm can scale to a
-  large number of samples > 100000, but does not scale in dimensionality:
-  n_features is expected to be lower than 100".
+In order to ease the reviewing process, we recommend that your contribution
+complies with the following rules before submitting a pull request. The
+**bolded** ones are especially important:
+
+1. **Make sure your code passes the tests**. The whole test suite can be run
+   with `pytest`, but it is usually not recommended since it takes a long
+   time. It is often enough to only run the test related to your changes:
+   for example, if you changed something in
+   `sklearn/linear_model/logistic.py`, running the following commands will
+   usually be enough:
+
+   - `pytest sklearn/linear_model/logistic.py` to make sure the doctest
+     examples are correct
+   - `pytest sklearn/linear_model/tests/test_logistic.py` to run the tests
+     specific to the file
+   - `pytest sklearn/linear_model` to test the whole `linear_model` module
+   - `pytest sklearn/doc/linear_model.rst` to make sure the user guide
+     examples are correct.
+
+   There may be other failing tests, but they will be caught by the CI so
+   you don't need to run the whole test suite locally. Read more in
+   `testing_coverage`_.
+
+2. **Make sure your code is properly commented and documented**, and **make
+   sure the documentation renders properly**. To build the documentation, please
+   refer to our `contribute_documentation`_ guidelines.
+
+3. **Tests are necessary for enhancements to be
+   accepted**. Bug-fixes or new features should be provided with
+   `non-regression tests
+   <https://en.wikipedia.org/wiki/Non-regression_testing>`_. These tests
+   verify the correct behavior of the fix or feature. In this manner, further
+   modifications on the code base are granted to be consistent with the
+   desired behavior. For the case of bug fixes, at the time of the PR, the
+   non-regression tests should fail for the code base in the master branch
+   and pass for the PR code.
+
+4. **Make sure that your PR does not add PEP8 violations**. Please run
+   `./build_tools/circle/flake8_diff.sh` or `make flake8-diff` on a Unix-like
+   system. You can also directly use `flake8 path_to_file`.
+
+5. Follow the `coding-guidelines`_ (see below).
+
+6. When applicable, use the validation tools and scripts in the
+   ``sklearn.utils`` submodule.  A list of utility routines available
+   for developers can be found in the :ref:`developers-utils` page.
+
+7. Give your pull request a helpful title that summarises what your
+   contribution does. In some cases "Fix <ISSUE TITLE>" is enough.
+   "Fix #<ISSUE NUMBER>" is not enough.
```

Comment:
This title will often become the commit message once merged so it should summarise your contribution for posterity.